network-store-model/src/main/java/com/powsybl/network/store/model/AttributesSpyer.java

```diff
@@ -0,0 +1,74 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.powsybl.network.store.model;
+
+import net.bytebuddy.ByteBuddy;
+import net.bytebuddy.implementation.MethodCall;
+import net.bytebuddy.implementation.SuperMethodCall;
+import net.bytebuddy.matcher.ElementMatchers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.EnumMap;
+
+/**
+ * @author Franck Lecuyer <franck.lecuyer at rte-france.com>
+ */
+public final class AttributesSpyer<T extends IdentifiableAttributes> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AttributesSpyer.class);
+
+    private static EnumMap<ResourceType, Class> attributesClasses = new EnumMap<>(ResourceType.class);
+
+    private T attributes;
+    private ResourceType resourceType;
+
+    private AttributesSpyer(T attributes, ResourceType resourceType) {
+        this.attributes = attributes;
+        this.resourceType = resourceType;
+    }
+
+    /*
+     * Construct an instance object of a subclass of T, dynamically generated with ByteBuddy, for each resource type,
+     * and with all setters intercepted to set after the dirty field to true
+     */
+    private T makeInstance() {
+        T instance = null;
+        try {
+            Class subClass;
+            if (!attributesClasses.containsKey(resourceType)) {
+                // create dynamically the subclass
+                subClass = new ByteBuddy()
+                        .subclass(attributes.getClass())
+                        .method(ElementMatchers.isSetter()
+                                .and(ElementMatchers.not(ElementMatchers.named("setDirty")))
+                                .and(ElementMatchers.not(ElementMatchers.named("setResource"))))
+                        .intercept(SuperMethodCall.INSTANCE.andThen(
+                                MethodCall.invoke(attributes.getClass().getMethod("setDirty", boolean.class))
+                                        .with(true)
+                        )).make()
+                        .load(attributes.getClass().getClassLoader())
+                        .getLoaded();
+                attributesClasses.put(resourceType, subClass);
+            } else {
+                subClass = attributesClasses.get(resourceType);
+            }
+
+            // create the new instance of this subclass
+            instance = (T) (subClass.getDeclaredConstructor(attributes.getClass()).newInstance(attributes));
+        } catch (NoSuchMethodException | InstantiationException | IllegalAccessException | InvocationTargetException e) {
+            LOGGER.error(e.toString(), e);
+        }
+
+        return instance;
+    }
+
+    public static<T extends IdentifiableAttributes> T create(T attributes, ResourceType resourceType) {
+        return (T) (new AttributesSpyer(attributes, resourceType).makeInstance());
```

Comment:
Is there a reason for having a newAttributeSpyer(a,b).makeInstance() instead of makeInstance(a,b) ?

Reply:
Changed